community/primitive-collections/src/test/java/org/neo4j/concurrent/AsyncEventsTest.java

```diff
@@ -0,0 +1,190 @@
+    {
+        EventConsumer consumer = new EventConsumer();
+
+        AsyncEvents<Event> asyncEvents = new AsyncEvents<>( consumer, AsyncEvents.Monitor.NONE );
+        executor.submit( asyncEvents );
+
+        asyncEvents.send( new Event() );
+
+        Thread processingThread = consumer.poll( 10, TimeUnit.SECONDS ).processedBy;
+        asyncEvents.shutdown();
+
+        assertThat( processingThread, is( not( Thread.currentThread() ) ) );
+    }
+
+    @Test( timeout = 10000 )
+    public void mustProcessEventsDirectlyWhenShutDown() throws Exception
+    {
+        EventConsumer consumer = new EventConsumer();
+
+        AsyncEvents<Event> asyncEvents = new AsyncEvents<>( consumer, AsyncEvents.Monitor.NONE );
+        executor.submit( asyncEvents );
+
+        asyncEvents.send( new Event() );
+        Thread threadForFirstEvent = consumer.poll( 10, TimeUnit.SECONDS ).processedBy;
+        asyncEvents.shutdown();
+
+        assertThat( threadForFirstEvent, is( not( Thread.currentThread() ) ) );
+
+        Thread threadForSubsequentEvents;
+        do
+        {
+            asyncEvents.send( new Event() );
+            threadForSubsequentEvents = consumer.poll( 10, TimeUnit.SECONDS ).processedBy;
+        }
+        while ( threadForSubsequentEvents != Thread.currentThread() );
+    }
+
+    @Test( timeout = 10000 )
+    public void concurrentlyPublishedEventsMustAllBeProcessed() throws Exception
+    {
+        EventConsumer consumer = new EventConsumer();
+        final CountDownLatch startLatch = new CountDownLatch( 1 );
+        final int threads = 10;
+        final int iterations = 10_000;
+        final AsyncEvents<Event> asyncEvents = new AsyncEvents<>( consumer, AsyncEvents.Monitor.NONE );
+        executor.submit( asyncEvents );
+
+        ExecutorService threadPool = Executors.newFixedThreadPool( threads );
+        Runnable runner = new Runnable()
+        {
+            @Override
+            public void run()
+            {
+                try
+                {
+                    startLatch.await();
+                }
+                catch ( InterruptedException e )
+                {
+                    throw new RuntimeException( e );
+                }
+
+                for ( int i = 0; i < iterations; i++ )
+                {
+                    asyncEvents.send( new Event() );
+                }
+            }
+        };
+        for ( int i = 0; i < threads; i++ )
+        {
+            threadPool.submit( runner );
+        }
+        startLatch.countDown();
+
+        Thread thisThread = Thread.currentThread();
+        int eventCount = threads * iterations;
+        try
+        {
+            for ( int i = 0; i < eventCount; i++ )
+            {
+                Event event = consumer.poll( 1, TimeUnit.SECONDS );
+                assertThat( event.processedBy, is( not( thisThread ) ) );
+            }
+        }
+        finally
+        {
+            asyncEvents.shutdown();
+        }
+    }
```

Comment:
In the other PR, I added a commit where I added a test for `awaitTermination`. That would be a good thing to include.